### PR TITLE
Alexnet nn.Graph version

### PR DESCRIPTION
**Pull request by strint**
alexnet net graph run
### 这个pr要解决的：
- [x] 搭Module模型，复用之前的
- [x] 单卡训练执行train
- [x] 单卡验证执行eval
- [x] loss 对齐，单卡graph  和 eager by @rentainhe 
- [x] 添加 infer 脚本
- [x] 把alexnet的dataloader改成nn.Module风格
- [x] 新增：把alexnet的dataloader Module 加入nn.Graph

### 后续pr要解决的：
- [ ] 部署到平台
- [ ] 低优先级，跑1小时预估下时间。跑完整的imagenet数据集，单机单卡，做完整任务的验证 by @rentainhe
- [ ] 单卡性能对比 dlperf
- [ ] 多卡和分布式版本
- [ ] 多卡分布式性能对比 dlperf
- [ ] 合并到 oneflow ci

### CODE REVIEW

File: alexnet/graph_train.sh [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+MOM=0.9
+EPOCH=2
+TRAIN_BATCH_SIZE=16
+VAL_BATCH_SIZE=16
+
+# LOAD PREVIOUS CHECKPOINT 
+# LOAD_CHECKPOINT=$CHECKPOINT_PATH/epoch_2_val_acc_0.111168
+
+export CUDA_VISIBLE_DEVICES=1
+<<<<<<< HEAD
```
Comment (MARD1NO) [inline alexnet/graph_train.sh:24]: 这个需要去除下，还有后面的那个 ====== >>>>>

Comment (strint) [reply alexnet/graph_train.sh:24]: @rentainhe 这个合并的冲突处理下

Comment (rentainhe) [reply alexnet/graph_train.sh:24]: 这边下一版提交会修改

**Comment by rentainhe** [comment]
## nn.Graph验证结果
- 模型：AlexNet
- 数据集：imagenette
- epochs：20
- optimizer：SGD
- learning rate：0.001
- momentum：0.9
- batchsize（train/val）：16/16

## Graph vs Eager
### Training Losses
losses

### Validation Accuracy
accuracy

File: alexnet/graph_train.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    parser.add_argument(
+        "--train_batch_size", type=int, default=32, help="train batch size"
+    )
+    parser.add_argument("--val_batch_size", type=int, default=32, help="val batch size")
+
+    return parser.parse_args()
+
+
+def main(args):
+    train_data_loader = OFRecordDataLoader(
```
Comment (strint) [inline alexnet/graph_train.py:40]: 这个为啥不是 nn.Module 呢 ？ from 成诚

Comment (Ldpe2G) [reply alexnet/graph_train.py:40]: 应该可以改成 module

**Comment by rentainhe** [comment]
如果需要复现AlexNet从头训练的结果，注意需要超参和原始paper对齐
- epoch = 90
- 初始学习率: 0.01
- optimizer: SGD（momentum=0.9, weight_decay=0.0005）
- schedule: StepLR（step_size=30,  gamma=0.1）
- train_batch_size = 128
- val_batch_size：看显存占用情况而定

**Comment by rentainhe** [comment]
## nn.Graph验证结果
- 模型：AlexNet
- 数据集：imagenette
- epochs：3
- optimizer：SGD
- learning rate：0.001
- momentum：0.9
- batchsize（train/val）：16/16

## 运行情况
显存占用：
Graph: 4M
Eager: 2M

## Graph vs Eager
### Training Losses
losses

### Validation Accuracy
acc

### Training Time (Epochs)
每个epoch训练所需要的时间比较
- 柱状图
Train_time_bar

- 折线图
train_time_epoch

### Training Time (Steps)
每个batch训练所需要的时间比较
training_time_step

### Evaluation Time (Epochs)
每个epoch测试所需要的时间比较
- 柱状图
Eval_time_bar

- 折线图
eval_time_epoch

File: alexnet/graph_train.sh [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+LEARNING_RATE=0.001
+MOM=0.9
+EPOCH=20
+TRAIN_BATCH_SIZE=16
+VAL_BATCH_SIZE=16
+
+# LOAD PREVIOUS CHECKPOINT 
+# LOAD_CHECKPOINT=$CHECKPOINT_PATH/epoch_2_val_acc_0.111168
+
+export CUDA_VISIBLE_DEVICES=1
```
Comment (Ldpe2G) [inline alexnet/graph_train.sh:23]: 这种环境变量最好还是不要放脚本里，直接在运行脚本的时候设置 
`CUDA_VISIBLE_DEVICES=xx bash graph_train.sh` ，
而且如果是只有单卡的用户，运行会报错，这里指定了卡1

File: alexnet/eager_train.sh [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    mkdir $CHECKPOINT_PATH
+fi
+
+LEARNING_RATE=0.001
+MOM=0.9
+EPOCH=90
+TRAIN_BATCH_SIZE=32
+VAL_BATCH_SIZE=32
+
+export CUDA_VISIBLE_DEVICES=2
```
Comment (Ldpe2G) [inline alexnet/eager_train.sh:20]: 这里一样

File: alexnet/eager_train.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        # save model after each epoch
+        print("***** Save Checkpoint *****")
+        save_path = os.path.join(args.save_checkpoint_path, "epoch_%d_val_acc_%f" % (epoch, accuracy))
+        save_checkpoint(model, save_path)
+        print("Save checkpoint to: ", save_path)
+        
+        # save best model
+        if best_acc < accuracy:
+            # TODO: 无法重复保存同一个名字的文件？我记得pytorch里是可以覆盖的
+            save_path = os.path.join(args.save_checkpoint_path, "best_model")
```
Comment (Ldpe2G) [inline alexnet/eager_train.py:223]: 我们是不会覆盖，需要手动删除目录
https://docs.python.org/3/library/shutil.html#shutil.rmtree

File: alexnet/eager_train.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+    #     batch_size=args.train_batch_size,
+    # )
+
+    # val_data_loader = OFRecordDataLoader(
+    #     ofrecord_root=args.ofrecord_path,
+    #     mode="validation",
+    #     dataset_size=50000,
+    #     batch_size=args.val_batch_size,
+    # )
+
```
Comment (Ldpe2G) [inline alexnet/eager_train.py:169]: 上面这段代码，做完实验记得删了

File: alexnet/eager_train.py (hunk 5 above)
Comment (rentainhe) [reply alexnet/eager_train.py:223]: 稍等，我这里也update一下

File: alexnet/eager_train.py (hunk 6 above)
Comment (rentainhe) [reply alexnet/eager_train.py:169]: OK，在修改ing

File: alexnet/check.sh [hunk 7]
```diff
@@ -13,14 +13,15 @@ fi
...
 LEARNING_RATE=0.001
 MOM=0.9
-EPOCH=20
+EPOCH=2
 TRAIN_BATCH_SIZE=16
 VAL_BATCH_SIZE=16
 
 # LOAD PREVIOUS CHECKPOINT 
 # LOAD_CHECKPOINT=$CHECKPOINT_PATH/epoch_2_val_acc_0.111168
+export GLOG_v=2
```
Comment (strint) [inline alexnet/check.sh:22]: 这是是debug的，需要去掉

File: alexnet/check.sh [hunk 8]
```diff
@@ -13,14 +13,15 @@ fi
 
 LEARNING_RATE=0.001
 MOM=0.9
-EPOCH=20
+EPOCH=2
```
Comment (strint) [inline alexnet/check.sh:16]: 恢复成20

File: alexnet/check.sh (hunk 7 above)
Comment (rentainhe) [reply alexnet/check.sh:22]: 好的

File: alexnet/check.sh (hunk 8 above)
Comment (rentainhe) [reply alexnet/check.sh:16]: OK

File: alexnet/check.sh [hunk 9]
```diff
@@ -20,7 +20,7 @@ VAL_BATCH_SIZE=16
 # LOAD PREVIOUS CHECKPOINT 
 # LOAD_CHECKPOINT=$CHECKPOINT_PATH/epoch_2_val_acc_0.111168
```
Comment (strint) [inline alexnet/check.sh:22]: 整理成这样的文件结构：
```
eager
    train.sh
    train.py
    infer.sh
    infer.py
graph
    train.sh
    train.py
    infer.sh
    infer.py
check
    check.sh
    check.py
```

File: alexnet/check_model.py [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+    print("***** Eager Training *****")
+    print("***** Preparing Dataloader *****")
+    eager_train_loader, eager_val_loader = build_loader(args)
+    print("***** Done *****")
+    trainer.eager_train(train_data_loader = eager_train_loader, 
+                        val_data_loader = eager_val_loader, 
+                        model = eager_model, 
+                        criterion = criterion, 
+                        optimizer = eager_optimizer)
+    save_result(trainer)
```
Comment (strint) [inline alexnet/check_model.py:330]: 行末尾增加一个空行

File: alexnet/check_model.py [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+        graph_model.parameters(), lr=args.learning_rate, momentum=args.mom
+    )
+
+    # criterion setup
+    criterion = flow.nn.CrossEntropyLoss()
+    criterion = criterion.to("cuda")
+
+    return eager_model, graph_model, eager_optimizer, graph_optimizer, criterion
+
+# def build_graph(args, graph_model, criterion, optimizer):
```
Comment (strint) [inline alexnet/check_model.py:84]: 废弃代码删掉

File: alexnet/check.sh (hunk 9 above)
Comment (rentainhe) [reply alexnet/check.sh:22]: OK

File: alexnet/check_model.py (hunk 10 above)
Comment (strint) [reply alexnet/check_model.py:330]: 这里只是保存了结果，但是文件的功能是做check。这里看能否做下程序的check，比如多轮迭代的loss差值的上限。

File: alexnet/eager_train.sh [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+python3 eager_train.py \
+    --save_checkpoint_path $CHECKPOINT_PATH \
+    --ofrecord_path $OFRECORD_PATH \
+    --learning_rate $LEARNING_RATE \
+    --mom $MOM \
+    --epochs $EPOCH \
+    --train_batch_size $TRAIN_BATCH_SIZE \
+    --val_batch_size $VAL_BATCH_SIZE \
+    # --load_checkpoint $LOAD_CHECKPOINT
```
Comment (strint) [inline alexnet/eager_train.sh:31]: 没用的话，就删掉吧

File: alexnet/check_model.py (hunk 10 above)
Comment (rentainhe) [reply alexnet/check_model.py:330]: 好的，我先修改一下

File: alexnet/graph_infer.sh [hunk 13]
```diff
@@ -0,0 +1,18 @@
+set -aux
+
+PRETRAIN_MODEL_PATH="./alexnet_oneflow_model"
+IMAGE_PATH="data/fish.jpg"
+# IMAGE_PATH="data/tiger.jpg"
+# IMAGE_PATH="data/ILSVRC2012_val_00020287.JPEG"
```
Comment (strint) [inline alexnet/graph_infer.sh:6]: 废弃的清理掉

File: alexnet/graph_train.sh [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+fi
+
+LEARNING_RATE=0.001
+MOM=0.9
+EPOCH=2
+TRAIN_BATCH_SIZE=16
+VAL_BATCH_SIZE=16
+
+# LOAD PREVIOUS CHECKPOINT 
+# LOAD_CHECKPOINT=$CHECKPOINT_PATH/epoch_2_val_acc_0.111168
```
Comment (strint) [inline alexnet/graph_train.sh:21]: 废弃代码

File: alexnet/model/__init__.py [hunk 15]
```diff
@@ -0,0 +1 @@
+from .alexnet import alexnet
```
Comment (strint) [inline alexnet/model/__init__.py:1]: 空行

Comment (rentainhe) [reply alexnet/model/__init__.py:1]: 好的

File: alexnet/README.md [hunk 16]
```diff
@@ -1,37 +1,58 @@
...
 ```
 
-### Run Oneflow Training script
+### 2. Run Oneflow Training Script
+#### Graph Training Scripts
+```bash
+bash graph/train.sh
+```
 
+#### Eager Training Scripts
```
Comment (strint) [inline alexnet/README.md:30]: eager在前，graph在后

File: alexnet/graph/train.py [hunk 17]
```diff
@@ -73,25 +76,54 @@ def main(args):
...
+    class AlexNetGraph(flow.nn.Graph):
+        def __init__(self):
+            super().__init__()
+            self.alexnet = alexnet_module
+            self.cross_entropy = of_cross_entropy
+            self.add_optimizer("sgd", of_sgd)
+            self.train_data_loader = train_data_loader
+        
+        def build(self):
+            image, label = self.train_data_loader()
```
Comment (strint) [inline alexnet/graph/train.py:88]: 这样写，现在可以正常执行不

Comment (rentainhe) [reply alexnet/graph/train.py:88]: 可以正常执行
acc

File: alexnet/README.md [hunk 18]
```diff
@@ -1,37 +1,58 @@
 # AlexNet
+Training Alexnet on [imagenette](https://github.com/fastai/imagenette) Dataset using OneFlow
 
-## Train on [imagenette](https://github.com/fastai/imagenette) Dataset
-
-### Prepare Traning Data And Pretrain Models
+## Usage
+### 0. Requirements
+Experiment environment:
```
Comment (Ldpe2G) [inline alexnet/README.md:6]: 这一节应该是和 requirements.txt 内容一致？

Comment (rentainhe) [reply alexnet/README.md:6]: 这个应该没什么问题吧。感觉正常一个models repo应该是这样写吧

**Review by Ldpe2G — APPROVED** [review]
